Firewall/v18

victorjulien opened the pull request:
SV_BRANCH=https://github.com/OISF/suricata-verify/pull/2380

First stab at allowing a firewall ruleset. Read the docs in the PR for an idea where this is going.

High level design, with (simplified) http example:
![FW rules per state - tables](https://github.com/user-attachments/assets/5ac75407-e282-4b19-b543-81ed726e9f99)

victorjulien commented on src/detect-engine-analyzer.c:
```diff
@@ -1929,3 +1959,221 @@ void EngineAnalysisRules(const DetectEngineCtx *de_ctx,
+        if (!AppProtoIsValid(a))
+            continue;
+
+        // HACK not all protocols have named states yet
```
should at least add support the built-in request_started, request_completed + same for response here

victorjulien commented on src/tx-bit.c:
```diff
@@ -0,0 +1,85 @@
+/* Copyright (C) 2014-2021 Open Information Security Foundation
```
2025

victorjulien commented on suricata.yaml.in:
```diff
@@ -2241,6 +2241,9 @@ napatech:
 
 default-rule-path: @e_defaultruledir@
 
+firewall-rule-files:
```
should be commented out

victorjulien commented on doc/userguide/firewall/firewall-example.rst:
```diff
@@ -0,0 +1,93 @@
+The HTTP rules need to ``accept`` each state::
+
+    # allow traffic before the request line is complete
+    accept:hook http1:request_not_started any any -> any any (sid:100;)
```
outdated

victorjulien commented on doc/userguide/rules/app-layer.rst:
```diff
@@ -97,3 +97,16 @@ applayer_proto_detection_skipped
+Examples::
+
+    app-layer-state:request_headers;
+    app-layer-state:>request_body;
```
need to add a full rule example

victorjulien commented on src/detect-app-layer-state.c:
```diff
@@ -0,0 +1,229 @@
+    g_applayer_state_list_id = DetectBufferTypeGetByName("app-layer-state");
+}
+
+static uint8_t DetectEngineAptStateInspect(DetectEngineCtx *de_ctx, DetectEngineThreadCtx *det_ctx,
```
can try make this a regular keyword now that the request_update logic is removed

victorjulien commented on src/detect-engine-alert.c:
```diff
@@ -208,6 +207,8 @@ static void PacketApplySignatureActions(Packet *p, const Signature *s, const Pac
         } else if (pa->action & (ACTION_ALERT | ACTION_CONFIG)) {
             // nothing to set in the packet
+        } else if (pa->action & (ACTION_ACCEPT)) {
+            // DEBUG_VALIDATE_BUG_ON((p->action & ACTION_ACCEPT) == 0); // should be unreachable
```
app accept doesn't set an explicit accept, but it might need to for cleaner logic

victorjulien commented on src/detect-engine-loader.c:
```diff
@@ -164,8 +169,18 @@ static int DetectLoadSigFile(DetectEngineCtx *de_ctx, const char *sig_file, int
+        else
+            sig = DetectEngineAppendSig(de_ctx, line);
         if (sig != NULL) {
+#if 0
```
remove

victorjulien commented on src/detect-engine-prefilter-common.c:
```diff
@@ -142,8 +142,9 @@ static int SetupEngineForPacketHeader(DetectEngineCtx *de_ctx, SigGroupHead *sgh
             ctx->sigs_cnt);
-    PrefilterAppendEngine(
-            de_ctx, sgh, Match, mask, ctx, PrefilterPacketHeaderFree, sigmatch_table[sm_type].name);
+    enum SignatureHookPkt hook = SIGNATURE_HOOK_PKT_NOT_SET; // TODO review
```
review

victorjulien commented on src/detect-engine-prefilter-common.c:
```diff
@@ -245,7 +246,8 @@ static int SetupEngineForPacketHeaderPrefilterPacketU8HashCtx(DetectEngineCtx *d
 
     if (cnt) {
-        PrefilterAppendEngine(de_ctx, sgh, Match, mask, ctx, PrefilterPacketU8HashCtxFree,
+        enum SignatureHookPkt hook = SIGNATURE_HOOK_PKT_NOT_SET; // TODO review
```
review

victorjulien commented on src/detect-engine-prefilter.c:
```diff
@@ -983,8 +1677,9 @@ int PrefilterGenericMpmPktRegister(DetectEngineCtx *de_ctx, SigGroupHead *sgh, M
     pectx->mpm_ctx = mpm_ctx;
     pectx->transforms = &mpm_reg->transforms;
 
+    enum SignatureHookPkt hook = SIGNATURE_HOOK_PKT_NOT_SET; // TODO review
```
review

victorjulien commented on src/detect-flowbits.c:
```diff
@@ -1332,8 +1332,9 @@ static int PrefilterSetupFlowbits(DetectEngineCtx *de_ctx, SigGroupHead *sgh)
     if (isset_ctx != NULL) {
-        PrefilterAppendEngine(de_ctx, sgh, PrefilterFlowbitMatch, SIG_MASK_REQUIRE_FLOW, isset_ctx,
-                PrefilterFlowbitFree, g_prefilter_flowbits_isset);
+        enum SignatureHookPkt hook = SIGNATURE_HOOK_PKT_NOT_SET; // TODO review
```
review

victorjulien commented on src/detect-lua.c:
```diff
@@ -913,76 +745,22 @@ static int DetectLuaSetup (DetectEngineCtx *de_ctx, Signature *s, const char *st
+    int list = DetectBufferGetActiveList(de_ctx, s);
+    SCLogDebug("buffer list %d -> %d", list, s->init_data->list);
+    if (list == -1 || (list == 0 && s->init_data->list == INT_MAX)) {
+        /* what needs to happen here is: we register to the rule hook, so e.g.
```
update/remove comment

victorjulien commented on src/detect-parse.c:
```diff
@@ -1258,12 +1520,85 @@ static uint8_t ActionStringToFlags(const char *action)
 
+    /* parse scope, if any */
+    if (o) {
+        uint8_t scope_flags = 0;
```
nit: not really a flag

victorjulien commented on src/detect-parse.c:
```diff
@@ -2098,6 +2438,17 @@ static void SigSetupPrefilter(DetectEngineCtx *de_ctx, Signature *s)
     SCReturn;
 }
 
+static bool DetectFilewallRuleValidate(const DetectEngineCtx *de_ctx, const Signature *s)
```
what is a filewall... wall of files?

victorjulien commented on src/detect-parse.c:
```diff
@@ -2098,6 +2438,17 @@ static void SigSetupPrefilter(DetectEngineCtx *de_ctx, Signature *s)
+static bool DetectFilewallRuleValidate(const DetectEngineCtx *de_ctx, const Signature *s)
+{
+    if (s->init_data->hook.type == SIGNATURE_HOOK_TYPE_NOT_SET) {
+        SCLogError("rule %u is loaded as a firewall rule, but does not specify and "
```
`an`

victorjulien commented on src/detect-parse.c:
```diff
@@ -2194,6 +2551,22 @@ static int SigValidate(DetectEngineCtx *de_ctx, Signature *s)
+                if (s->init_data->hook.type == SIGNATURE_HOOK_TYPE_APP) {
+                    if ((s->flags & SIG_FLAG_TOSERVER) && (app->dir == 0) &&
+                            app->progress != s->init_data->hook.t.app.app_progress) {
+                        SCLogError("engine progress value %d doesn't match hook %u", app->progress,
```
can use named representation of the hook for better error msg

victorjulien commented on src/detect-parse.c:
```diff
@@ -2352,15 +2725,20 @@ static int SigValidate(DetectEngineCtx *de_ctx, Signature *s)
+    if (firewall_rule) {
+        sig->init_data->firewall_rule = true;
+        sig->flags |= SIG_FLAG_FIREWALL;
+        de_ctx->flags |= DE_HAS_FIREWALL;
```
should probably set this if firewall rule loading is enabled instead

victorjulien commented on src/detect.c:
```diff
@@ -139,9 +132,20 @@ static void DetectRun(ThreadVars *th_v,
+     *
+     * accept:hook won't have set the pkt_policy, so we simply continue.
+     *
+     * TODO what about app state progression, cleanup and such? */
```
review

victorjulien commented on src/detect.c:
```diff
@@ -805,6 +650,24 @@ static inline void DetectRulePacketRules(
+            if (p->flags & PKT_NOPACKET_INSPECTION ||
+                    (pflow != NULL && pflow->flags & (FLOW_ACTION_PASS))) {
+                SCLogDebug("skipping firewall rule %u", s->id);
+                break_out_of_packet_filter = true;
```
should not be needed to set this? Current rule is not a fw rule, so we're already beyond packet:filter

victorjulien commented on src/detect.c:
```diff
@@ -926,12 +789,63 @@ static inline void DetectRulePacketRules(
             }
         }
+
+        // TODO this can move into a per FW rule post-Match func
```
or at least a helper func

victorjulien commented on src/detect.c:
```diff
@@ -1534,6 +1462,76 @@ static inline void RuleMatchCandidateMergeStateRules(
 
+/** \internal
+ *  Check and update firewall rules state.
+ *  \param skip_fw_hook bool to indicate firewall rules for state `skip_before_progress` should
```
formatting here is clang-format's doing :-/

victorjulien commented on src/detect.c:
```diff
@@ -1534,6 +1462,76 @@ static inline void RuleMatchCandidateMergeStateRules(
+         * progress beyond our progress + 1, it means the next progress has no rules and needs
+         * the default policy applied. But only after we evaluate the current rule first, as
+         * that may override it.
+         * TODO should we do this after dedup below? */
```
remove todo comment

victorjulien commented on src/detect.c:
```diff
@@ -1721,6 +1766,58 @@ static void DetectRunTx(ThreadVars *tv,
+                        } else if (as == ACTION_SCOPE_FLOW) {
+                            break_out_of_app_filter = true;
+
+                            // TODO this should be postponed to FinalizeAlerts
```
do this

victorjulien commented on src/detect.c:
```diff
@@ -1790,6 +1893,14 @@ static void DetectRunTx(ThreadVars *tv,
+
+    /* apply default policy if there were txs to inspect, we have fw rules and non of the rules
+     * applied a policy. */
+    if (tx_inspected && have_fw_rules && !fw_verdict) {
```
fw_verdict may be set in tx 0, but then tx 1 might not have eval'd any rules in case of prefiltered rules? Then we should drop for tx 1, but since tx 0 already set this to true we won't? Need a test for this.

victorjulien replied:
maybe fw_verdict and tx_inspected both should be counters, and if they aren't equal at the end, we drop

victorjulien commented on src/detect-engine-alert.c:
```diff
@@ -421,6 +432,11 @@ static inline void PacketAlertFinalizeProcessQueue(
                 break;
             }
+
+            // TODO we can also drop if alert is suppressed, right?
```
review this

victorjulien commented on src/detect-engine-sigorder.c:
```diff
@@ -723,6 +723,50 @@ static int SCSigOrderByPriorityCompare(SCSigSignatureWrapper *sw1,
     return 0;
 }
 
+static int SCSigOrderByIId(SCSigSignatureWrapper *sw1, SCSigSignatureWrapper *sw2)
```
pondering renaming Signature::num to Signature::iid, which is where this name comes from. Should probably rename to something clearer

jufajardini replied:
Are Signature nums unique? They are, right? Could it be Signature::uid?

victorjulien commented on src/detect-engine-sigorder.c:
```diff
@@ -723,6 +723,50 @@ static int SCSigOrderByPriorityCompare(SCSigSignatureWrapper *sw1,
+}
+
+/* sort by:
+ * alproto, progress, iid
```
direction

victorjulien commented:
Some generic todos:
- block thresholds for fw rules for now
- block frame rules for now

jasonish commented on doc/userguide/firewall/firewall-example.rst:
```diff
@@ -0,0 +1,93 @@
+Firewall Ruleset Examples
```
I think some very sipmle examples would be useful as well.
- allow all HTTP, tls, ssh, dns
- drop all

Or the better common case
- allow everything outgoing stateful
- only allow incoming ssh, HTTP, https

jasonish commented on etc/schema.json:
```diff
@@ -66,6 +66,12 @@
         "payload": {
             "type": "string"
         },
+        "ts_progress": {
```
It looks like its used to convery application-layer state? `app_state_ts`, `app_state_tc`?

Also, new fields should get a description.

suricata-qa commented:
Information: QA ran without warnings. 

Pipeline 25520

jasonish commented on doc/userguide/firewall/firewall-design.rst:
```diff
@@ -0,0 +1,182 @@
+
+* ``packet`` accept this packet
+* ``flow`` accept the rest of the packets in this flow
+* ``hook`` accept rules for the current hook/state, evaluate the next tables
```
Is `state` a better name for this? The state of a connection or protocol makes sense and is in-line with some current lingo.  Plus below we describe hook as state.. For example: `Explicit rule hook (states)`.

jasonish commented:
Some observations...

### The default "drop" policy is not consistent.

If I use an empty firewall rule file, nothing is dropped.

If I use a single rule:
```
accept:flow icmp:all $HOME_NET any -> any any (sid:2;)
```
ICMP works, but I can still connect to any TCP port.

But a single rule like:
```
accept:flow tcp:all any any -> any 53 (sid:6;)
```
will block all tcp except port 53, but allow all ICMP and UDP.

So it appears you need one accept rule on a protocol for the drop by default policy?

jasonish commented:
Another one... As the drop by default wasn't working, I tried a ruleset like:
```
# Default drop.
drop:packet ip:all any any -> any any (sid: 4;)

# Allow outgoing ICMP and replies.
accept:flow icmp:all $HOME_NET any -> any any (sid:2;)

# Allow outgoing DNS and replies.
accept:flow udp:all $HOME_NET any -> any 53 (sid:3;)
accept:flow tcp:all $HOME_NET any -> any 53 (sid:5;)
```

If I start a ping, the first ping is lost (sequence number 1), but the second echo request goes out and responses on received.  With the drop at the top, I wouldn't expect this?

Likewise for TCP, the first syn is dropped, but the retransmission goes through.  So I can do TCP DNS requests.

I can't get a UDP packet out.

victorjulien commented:
> Some observations...
> ### The default "drop" policy is not consistent.
> 
> If I use an empty firewall rule file, nothing is dropped.
> 
> If I use a single rule:
> 
> ```
> accept:flow icmp:all $HOME_NET any -> any any (sid:2;)
> ```
> 
> ICMP works, but I can still connect to any TCP port.
> 
> But a single rule like:
> 
> ```
> accept:flow tcp:all any any -> any 53 (sid:6;)
> ```
> 
> will block all tcp except port 53, but allow all ICMP and UDP.
> 
> So it appears you need one accept rule on a protocol for the drop by default policy?

Fixing this, it turns out we defaulted to "accept" if no rule group existed.

victorjulien commented:
Pushed a rework of the default action handling, now requiring an explicit ACCEPT for every packet internally. Should hopefully solve the issues @jasonish saw.

Code still needs more cleanup.

jasonish commented:
> Pushed a rework of the default action handling, now requiring an explicit ACCEPT for every packet internally. Should hopefully solve the issues @jasonish saw.
> 
> Code still needs more cleanup.

Can confirm that drop by default now seems to work.

The retransmission issue still exists for a ruleset with a drop at the top though:
```
# Default drop.
drop:packet ip:all any any -> any any (sid: 4;)

# Allow outgoing ICMP and replies.
accept:flow icmp:all $HOME_NET any -> any any (sid:2;)

# Allow outgoing DNS and replies.
accept:flow udp:all $HOME_NET any -> any 53 (sid:3;)
accept:flow tcp:all $HOME_NET any -> any 53 (sid:5;)
```

victorjulien commented:
> > Pushed a rework of the default action handling, now requiring an explicit ACCEPT for every packet internally. Should hopefully solve the issues @jasonish saw.
> > Code still needs more cleanup.
> 
> Can confirm that drop by default now seems to work.
> 
> The retransmission issue still exists for a ruleset with a drop at the top though:
> 
> ```
> # Default drop.
> drop:packet ip:all any any -> any any (sid: 4;)
> 
> # Allow outgoing ICMP and replies.
> accept:flow icmp:all $HOME_NET any -> any any (sid:2;)
> 
> # Allow outgoing DNS and replies.
> accept:flow udp:all $HOME_NET any -> any 53 (sid:3;)
> accept:flow tcp:all $HOME_NET any -> any 53 (sid:5;)
> ```

Not able to reproduce this currently. What happens if you remove sid 4?

jasonish commented:
> > > Pushed a rework of the default action handling, now requiring an explicit ACCEPT for every packet internally. Should hopefully solve the issues @jasonish saw.
> > > Code still needs more cleanup.
> > 
> > 
> > Can confirm that drop by default now seems to work.
> > The retransmission issue still exists for a ruleset with a drop at the top though:
> > ```
> > # Default drop.
> > drop:packet ip:all any any -> any any (sid: 4;)
> > 
> > # Allow outgoing ICMP and replies.
> > accept:flow icmp:all $HOME_NET any -> any any (sid:2;)
> > 
> > # Allow outgoing DNS and replies.
> > accept:flow udp:all $HOME_NET any -> any 53 (sid:3;)
> > accept:flow tcp:all $HOME_NET any -> any 53 (sid:5;)
> > ```
> 
> Not able to reproduce this currently. What happens if you remove sid 4?

If I remove SID 4, the TCP DNS request goes through as expected.

But when I keep SID 4, the first SYN gets dropped, the retransmission does not, and the TCP flow is accepted:
```
{"timestamp":"2025-04-04T14:52:52.049037-0600","flow_id":1336515275367512,"event_type":"alert","src_ip":"10.88.0.3","src_port":44975,"dest_ip":"1.1.1.1","dest_port":53,"proto":"TCP","pkt_src":"wire/pcap","alert":{"action":"blocked","gid":1,"signature_id":4,"rev":0,"signature":"","category":"","severity":3},"direction":"to_server","flow":{"pkts_toserver":1,"pkts_toclient":0,"bytes_toserver":60,"bytes_toclient":0,"start":"2025-04-04T14:52:52.049037-0600","src_ip":"10.88.0.3","dest_ip":"1.1.1.1","src_port":44975,"dest_port":53}}
{"timestamp":"2025-04-04T14:52:53.088604-0600","flow_id":1336515275367512,"event_type":"dns","src_ip":"10.88.0.3","src_port":44975,"dest_ip":"1.1.1.1","dest_port":53,"proto":"TCP","pkt_src":"wire/pcap","dns":{"version":3,"type":"request","tx_id":0,"id":10403,"flags":"120","rd":true,"opcode":0,"rcode":"NOERROR","queries":[{"rrname":"codemonkey.net","rrtype":"A"}],"additionals":[{"rrname":"","rrtype":"OPT","ttl":0,"opt":[{"code":10,"data":"b4cf4c6f3a88e3a7"}]}]}}
{"timestamp":"2025-04-04T14:52:53.153034-0600","flow_id":1336515275367512,"event_type":"dns","src_ip":"10.88.0.3","src_port":44975,"dest_ip":"1.1.1.1","dest_port":53,"proto":"TCP","pkt_src":"wire/pcap","dns":{"version":3,"type":"response","tx_id":1,"id":10403,"flags":"8180","qr":true,"rd":true,"ra":true,"opcode":0,"rcode":"NOERROR","queries":[{"rrname":"codemonkey.net","rrtype":"A"}],"answers":[{"rrname":"codemonkey.net","rrtype":"A","ttl":600,"rdata":"152.53.82.239"}],"grouped":{"A":["152.53.82.239"]}}}
{"timestamp":"2025-04-04T14:52:55.227145-0600","flow_id":1336515275367512,"event_type":"flow","src_ip":"10.88.0.3","src_port":44975,"dest_ip":"1.1.1.1","dest_port":53,"proto":"TCP","app_proto":"dns","flow":{"pkts_toserver":7,"pkts_toclient":4,"bytes_toserver":437,"bytes_toclient":277,"start":"2025-04-04T14:52:52.049037-0600","end":"2025-04-04T14:52:53.196176-0600","age":1,"state":"closed","reason":"shutdown","alerted":true,"action":"accept"},"tcp":{"tcp_flags":"1b","tcp_flags_ts":"1b","tcp_flags_tc":"1b","syn":true,"fin":true,"psh":true,"ack":true,"state":"closed","ts_max_regions":1,"tc_max_regions":1}}
```

suricata-qa commented:
WARNING:
 
 | field | baseline | test | % |
|--|--|--|--| 
| | SURI_TLPW2_single_stats_chk | | |
| .detect.alerts_suppressed | 7864 | 8204809 | 104333.79% |
| | SURI_TLPW2_autofp_stats_chk | | |
| .detect.alerts_suppressed | 7864 | 8204811 | 104333.81% |
| | SURI_TLPW1_stats_chk | | |
| .detect.alerts_suppressed | 58568 | 5570870 | 9511.8% |
| | SURI_TLPR1_stats_chk | | |
| .detect.alerts_suppressed | 1489486 | 151543866 | 10174.24% |
| | IPS_AFP_stats_chk | | |
| .detect.alert_queue_overflow | 0 | 4320 | - |
| .detect.alerts_suppressed | 597240 | 282333144 | 47272.98% |
| | TREX_GENERIC_stats_chk | | |
| .detect.alerts_suppressed | 934036 | 275698062 | 29516.86% | 

Pipeline 25546

suricata-qa commented:
Information: QA ran without warnings. 

Pipeline 25549

suricata-qa commented:
Information: QA ran without warnings. 

Pipeline 25551

suricata-qa commented:
Information: QA ran without warnings. 

Pipeline 25554

victorjulien commented:
Passed my QA. Ran this PR with SV PR https://github.com/OISF/suricata-verify/pull/2380. Local pipeline 5344, run 813.

victorjulien commented:
Replaced by #12962. Main issue here was the retransmission fail open case. I've changed the default drop for the app:* tables to `drop:flow` for this. This matches the IPS drop logic, where in app we drop:flow. This is needed as the matching state gets cleaned up already, so a retransmission can't repeat that.
